### DISCUSSION TITLE
When using --upload-only files which only exist in OneDrive will NOT be deleted

### DISCUSSION THREAD

**Discussion by FlynnLG** [abraunegg/onedrive#2274] [Q&A] [answered]
Hello!

Before I start: I really like the OneDrive client for Linux! But I need some help for solving a problem when uploading files to OneDrive.

When I execute `onedrive --synchronize --upload-only` I expect the client to upload all files from my local disk to OneDrive and delete files on OneDrive which are not present on my local disk. Sadly the client does not delete files which are only on OneDrive (not on my local disk), but does correctly upload everything and does not download anything.

For example:
When I create a directory via OneDrive website and then run `onedrive --synchronize --upload-only`:
```
root@localhost:~/OneDrive# onedrive --synchronize --upload-only
Configuration file successfully loaded
Configuring Global Azure AD Endpoints
User Configured Rate Limit: 5242880
Initializing the Synchronization Engine ...
Syncing changes from local path only - NOT syncing data changes from OneDrive ...
Uploading differences of ~/OneDrive
Uploading new items of ~/OneDrive
Sync with OneDrive is complete
```
The client does not delete the directory I just created via the OneDrive website.

Anyone know why this might occur and/or what I did wrong?
Thanks in advance!
Best regards,
FlynnLG.

**Comment by abraunegg** [comment, marked as answer]
@FlynnLG 
> The client does not delete the directory I just created via the OneDrive website.

The client is operating correctly.

The client, when using `--upload-only`, is using your local directory as the source of truth, and anything 'local' that is not online, if it was synced prior will be removed.

The reason *why* your online created folder is not being removed, is because the client is *never* requesting what exists online - thats the nature of `--upload-only` .. your only uploading data to OneDrive - not requesting what exists there. As such there is no way for the client to understand / determine what is 'online' when using `--upload-only` - hence no delete.

~~Now - if you were to not use `--upload-only`, but use `--synchronize --local-first` - you potentially will have a different outcome, namely, the remote is synced with the local state, and anything remote, that is not local *should* be removed.~~

~~Test with `--dry-run` to validate this.~~

**Edit:**
I have tested `--synchronize --local-first` and the remote created folder is created locally, which, when I look at that use-case, is also correct. 

Thus - essentially what you are seeing is a *quirk* of using the client - as `--upload-only` never requests the online objects - as that is the nature of upload only.

**Reply by FlynnLG** [reply]
Hello @abraunegg!
Thanks for your answer!
Never had any doubt that the client operates correctly. Was pretty sure that I did miss something.

Okay! Thanks for your explanation. Now I understand the process behind `--upload-only`.

I tried `onedrive --synchronize --local-first --dry-run` and got the following output:
```
root@localhost:~/OneDrive# onedrive --synchronize --local-first --dry-run
Configuration file successfully loaded
DRY-RUN Configured. Output below shows what 'would' have occurred.
Configuring Global Azure AD Endpoints
User Configured Rate Limit: 5242880
Initializing the Synchronization Engine ...
Syncing changes from local path first before downloading changes from OneDrive ...
Uploading differences of ~/OneDrive
Uploading new items of ~/OneDrive
Creating local directory: testing
Sync with OneDrive is complete
```
It seems that the client tries to create the directory on my local disk and does not delete it online, although specifying `--local-first`.

**Reply by FlynnLG** [reply]
I read your edit and it seems there is no way around this "problem"?

**Reply by abraunegg** [reply]
> It seems that the client tries to create the directory on my local disk and does not delete it online, although specifying `--local-first`

Correct ... it was in incorrect suggestion by me as an alternative to try.

**Reply by abraunegg** [reply]
> I read your edit and it seems there is no way around this "problem"?

Correct - and it is not really a 'problem' or 'bug' either ... the client only knows about what 'it' has uploaded .. because of the use of 'upload only' .... anything done outside the client, is 'invisible' .... thus, cannot be impacted / changed / deleted without some 'other' operation request to 'cleanup online' - which means 'downloading' at least the whole online state metadata - which defeats the purpose of 'upload only' ....

**Reply by FlynnLG** [reply]
Okay! I unterstand that it's not possible to delete files created online - if not downloaded before - because that's how the OneDrive API works.

Thank you for helping me and keep maintaining the OneDrive client as good as it's now! :)

**Reply by abraunegg** [reply]
Thanks for the feedback :)
